### ISSUE TITLE
BKD-1a — Pin the canonical weight layout of the symmetric contraction

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1564]
**Depends on:** CORE-4 (#1558), BKD-1 (#1560), BKD-3 (#1567) · **Blocks:** FM-00 (#1565)

**Context:** One canonical weight layout is what lets a single checkpoint load into any backend without the five legacy weight-conversion CLIs. For the symmetric contraction, the layout is not an open design question — the measurements settled it — but it must be *written down and locked*, because it is the on-disk file format: nothing is serialized to the neutral format (FM-00 (#1565)) before this lands. Two measured facts collapse what used to look like a "nested vs fused parametrization" choice. First, the projection matrix between the two historical parametrizations is a **permutation times a diagonal** — exactly one non-zero per row and per column, with scales that are square roots of small integers (see the table below). Second, the nesting is a *view*: the legacy converter `mace/cli/convert_e3nn_cueq.py:118-127` **concatenates** the per-`(irrep_out, ν)` tensors along the path axis into one flat `[Z, A, mul]` tensor *before* projecting (`:143-150`), and the reverse converter (`mace/cli/convert_cueq_e3nn.py:119-159`) projects and then **splits** — so the per-path tensors are contiguous slices of one flat tensor. Therefore the canonical form is that flat `[Z, A, mul]` tensor, over the reduced Clebsch-Gordan basis, in a path order and per-path normalization that **MACE owns and pins** — not cueq's layout, and not an artefact of the nested split. CORE-4 (#1558) computes and pins the order and normalization; this ticket writes the convention down in `mace_core.clebsch_gordan`, golden-locks the serialization, and implements both boundary conversions.

**Interface & constraints:**

1. **Measured structure of the parametrization map** (production grid; the map from the flat-canonical form to the legacy nested form):

| `irreps_in` | `irreps_out` | correlation | `proj.shape` | permutation? | distinct scales |
|---|---|---|---|---|---|
| `0e+1o` | `0e+1o` | 3 | (9, 9) | yes | 1, √2 |
| `0e+1o+2e` | `0e+1o` | 3 | (17, 17) | yes | 1, √2, √3, √5 |
| `0e+1o+2e+3o` | `0e+1o` | 3 | (29, 29) | yes | 1, √3, 2, 2√2, 2√3 |
| `0e+1o+2e+3o` | `0e+1o+2e` | 3 | (49, 49) | yes | 1, √3, 2, √5, 2√2 |
| `0e+1o+2e` | `0e+1o` | 4 | (35, 35) | yes | 1, √2, √3, √5, 3 |

`proj = Permutation · Diagonal` in every case. The two historical forms differ by path order and per-path normalization, nothing more.

2. **The descriptor owns the layout.** The canonical path order and normalization for a given `(irreps_in, irreps_out, correlation)` are emitted alongside the reduced basis by `mace_core.clebsch_gordan`, so the op descriptor owns the layout exactly as it owns `weight_numel` (in BKD-1 (#1560)'s kernel contract, `weight_numel` and the weight order are properties of the descriptor, fixed by the canonical instruction set and identical across backends — never negotiated per backend).

3. **Null paths are stored as explicit zeros**, with a reachability mask in the descriptor. Legacy detects a null slot by comparing the whole `U_matrix` against zero (`mace/modules/symmetric_contraction.py:118`) and then replaces the parameter with an `EmptyParam` — a `torch.nn.Parameter` subclass that zeroes its data and refuses `requires_grad_` (`:269-278`) — while registering a companion `weights_{idx}_zeroed` / `weights_max_zeroed` boolean buffer (`:218-238`). A `(irrep_out, ν)` slot that is null this way (the output irrep is unreachable with ν factors — e.g. for `irreps_in = 0e+1o`, output `3o` is null at ν=1 and ν=2 and carries one path at ν=3) still occupies its slice in the canonical tensor. This keeps `weight_numel` a pure function of `(irreps_in, irreps_out, correlation)` and keeps the scaled permutation to any backend's internal order well-defined; a backend may drop null paths internally behind `load_canonical` without the file format knowing. Note which legacy branch does which: the **reduced** path takes the unfiltered concatenation, zeroed slots included (`mace/cli/convert_e3nn_cueq.py:117-127`), while the full-basis path filters on the `*_zeroed` buffers (`:128-142`). Canonical follows the reduced branch — zeros are stored.

4. **Boundary-only conversion.** `to_canonical`/`load_canonical` run only at the checkpoint save/load boundary, never in `forward`. For the reference backend they are `cat`/`split` along the path axis — views, no copy, byte-identity round trip. For the cueq backend, `load_canonical` is an index permutation plus a scale vector, derived once at build time (the cueq backend has `cuequivariance` present by definition) and cached in the op. If cueq changes its internal segmentation in a future release, only that index buffer changes — the file format does not.

5. **The convention is the file format.** The serialized spec carries a `schema_version` recording the pinned order and normalization, so a checkpoint is self-describing; changing the convention later is a `schema_version` bump with an in-repo exact conversion, never a silent reinterpretation.

**Task:**
1. In `mace_core.clebsch_gordan`, define and document the canonical path order and the per-path normalization for a given `(irreps_in, irreps_out, correlation)`. Emit them alongside the reduced basis, so the descriptor owns the layout exactly as it owns `weight_numel`.
2. Reference backend: `to_canonical` = `cat` along the path axis, `load_canonical` = `split`. Views, no copy. The Horner cascade is unchanged.
3. cueq backend: derive its index permutation and scale vector **once at build time** and cache them in the op. No projection in `forward`.
4. Golden-lock the convention: a committed reference recording the path order and each path's scale for a fixed `(irreps_in, irreps_out, correlation)` grid.
5. Conformance test `test_backend_reorder_is_scaled_permutation`: assert each backend's canonical→internal map has one non-zero per row and column. A future cueq whose segmentation made it dense fails loudly instead of silently reinterpreting weights.

**Out of scope:** any `SegmentedPolynomial` evaluator — not needed: the container is a view and the reference keeps its Horner cascade. A fused *reference* evaluator is a separate question, owned by BKD-1 (#1560)'s optional `interaction_layer` span factory, and the measurements argue against it: benchmarked pure-torch on CPU at fp64 (same reduced basis, 2 000 nodes, 128 channels, 89 elements, `correlation=3`), the Horner cascade beats a naive fused evaluator at `lmax≥2` (28.80 ms vs 40.68 ms at `lmax=2`/`0e`; 67.36 ms vs 188.06 ms at `lmax=3`/`0e`; the fused form wins only at `lmax=1`, 5.25 ms vs 8.00 ms). Fused evaluation's real advantage is memory-traffic fusion inside one GPU kernel, which the accelerated path gets under any canonical choice.

**Acceptance criteria:**
- [ ] Reference `to_canonical`/`load_canonical` are byte-identity round trips.
- [ ] A checkpoint written by the reference loads into the cueq backend and reproduces its outputs at fp64 tolerance.
- [ ] The path-order/per-path-scale golden is committed.
- [ ] `test_backend_reorder_is_scaled_permutation` is green for every registered backend.
- [ ] Null paths occupy their canonical slice as explicit zeros, and `weight_numel` is reproducible from `(irreps_in, irreps_out, correlation)` alone — no `EmptyParam`-style sentinel and no `*_zeroed` companion buffer survives into the file format.
- [ ] `mace_core` imports no `cuequivariance`.

**Verify:**
```bash
python -m pytest packages/mace-core/tests/test_canonical_layout.py -v
python -m pytest packages/mace-torch/tests/backends/test_weight_roundtrip.py -v
```
**Review focus:** the normalization convention — it is the file format, and changing it later is a `schema_version` bump.

---
